server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OffHeapStarTreeBuilder.java

```diff
@@ -0,0 +1,896 @@
+     * @param metricReaders    List of docValues readers to read metrics from the segment
+     * @return Iterator of star-tree documents
+     */
+    @Override
+    public Iterator<StarTreeDocument> sortAndAggregateSegmentDocuments(
+        SequentialDocValuesIterator[] dimensionReaders,
+        List<SequentialDocValuesIterator> metricReaders
+    ) throws IOException {
+        // Write all dimensions for segment documents into the buffer, and sort all documents using an int
+        // array
+        int documentBytesLength = 0;
+        int[] sortedDocIds = new int[totalSegmentDocs];
+        for (int i = 0; i < totalSegmentDocs; i++) {
+            sortedDocIds[i] = i;
+        }
+
+        try {
+            for (int i = 0; i < totalSegmentDocs; i++) {
+                StarTreeDocument document = getSegmentStarTreeDocument(i, dimensionReaders, metricReaders);
+                documentBytesLength = writeStarTreeDocument(document, segmentDocsFileOutput, false);
+            }
+        } finally {
+            segmentDocsFileOutput.close();
+        }
+
+        // Create an iterator for aggregated documents
+        return sortAndReduceDocuments(sortedDocIds, totalSegmentDocs, documentBytesLength);
+    }
+
+    /**
+     * Generates a star-tree for a given star-node
+     *
+     * @param startDocId  Start document id in the star-tree
+     * @param endDocId    End document id (exclusive) in the star-tree
+     * @param dimensionId Dimension id of the star-node
+     * @return iterator for star-tree documents of star-node
+     * @throws IOException throws when unable to generate star-tree for star-node
+     */
+    @Override
+    public Iterator<StarTreeDocument> generateStarTreeDocumentsForStarNode(int startDocId, int endDocId, int dimensionId)
+        throws IOException {
+        // End doc id is not inclusive but start doc is inclusive
+        // Hence we need to check if buffer is readable till endDocId - 1
+        ensureDocumentReadable(endDocId - 1);
+
+        // Sort all documents using an int array
+        int numDocs = endDocId - startDocId;
+        int[] sortedDocIds = new int[numDocs];
+        for (int i = 0; i < numDocs; i++) {
+            sortedDocIds[i] = startDocId + i;
+        }
+        new IntroSorter() {
+            private long[] dimensions;
+
+            @Override
+            protected void swap(int i, int j) {
+                int temp = sortedDocIds[i];
+                sortedDocIds[i] = sortedDocIds[j];
+                sortedDocIds[j] = temp;
+            }
+
+            @Override
+            protected void setPivot(int i) {
+                long offset = starTreeDocumentOffsets.get(sortedDocIds[i]);
+                dimensions = new long[starTreeField.getDimensionsOrder().size()];
+                try {
+                    for (int j = dimensionId + 1; j < dimensions.length; j++) {
+                        dimensions[j] = starTreeDocsFileRandomInput.readLong(offset + (long) j * Long.BYTES);
```

Comment:
Should we not use getDimValue here? What if we compare actual 0 with null?

Reply:
Good catch. Added a test case for this as well.